### PR TITLE
Adding ocp actuator

### PR DESCRIPTION
**Pull request by bardielle**

### CODE REVIEW

File: pkg/ocp/actuator.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+// Create the credentials.
+func (a *OCPActuator) Create(ctx context.Context, cr *minterv1.CredentialsRequest) error {
+	logger := a.getLogger(cr)
+	logger.Debug("running create")
+	return a.createSecret(ctx, cr, logger)
+}
+
+// Update the credentials to the provided definition.
+func (a *OCPActuator) Update(ctx context.Context, cr *minterv1.CredentialsRequest) error {
```
Comment (nirarg) [inline pkg/ocp/actuator.go:82]: You do here only create if doesn't exists, what about update existing secret?

Comment (bardielle) [reply pkg/ocp/actuator.go:82]: Why does it need to be updated? if the key exist it exist. if no it's not and need to create
As I said - in the future we will have expiration kubeconfig and if the kubeconfig expired exists will return false and then the reconcile will call to create. 
There isn't nothing to update here... 
This is the same behavior as was implemeted in other platforms

**Review by nirarg — APPROVED** [review]

File: pkg/ocp/actuator.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+			return err
+		}
+	}
+
+	return nil
+}
+
+// GetCredentialsRootSecretLocation returns the namespace and name where the parent credentials secret is stored.
+func (a *OCPActuator) GetCredentialsRootSecretLocation() types.NamespacedName {
+	return types.NamespacedName{Namespace: constants.CloudCredSecretNamespace, Name: constants.KubevirtCloudCredSecretName}
```
Comment (bardielle) [inline pkg/ocp/actuator.go:109]: @nirarg  I added that function as part of the commit -
Please re-review

File: pkg/ocp/actuator.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+func (a *OCPActuator) createSecret(ctx context.Context, cr *minterv1.CredentialsRequest, logger log.FieldLogger) error {
+	sLog := logger.WithFields(log.Fields{
+		"targetSecret": fmt.Sprintf("%s/%s", cr.Spec.SecretRef.Namespace, cr.Spec.SecretRef.Name),
+		"cr":           fmt.Sprintf("%s/%s", cr.Namespace, cr.Name),
+	})
+
+	// get the secret of the kubevirt credentials
+	kubevirtCredentialsSecret := &corev1.Secret{}
+	if err := a.Client.Get(ctx, a.GetCredentialsRootSecretLocation(), kubevirtCredentialsSecret); err != nil {
```
Comment (bardielle) [inline pkg/ocp/actuator.go:144]: I changed the get secret to use the function that I added above
@nirarg  - Please re-review

File: pkg/operator/constants/constants.go [hunk 4]
```diff
@@ -94,6 +94,10 @@ const (
 	// VSphereCloudCredSecretName is the name of the secret where credentials
 	// for vSphere are stored.
 	VSphereCloudCredSecretName = "vsphere-creds"
+
+	// KubevirtCloudCredSecretName is the name of the secret where credentials
+	// for Kubevirt are stored.
+	KubevirtCloudCredSecretName = "kubevirt-credentials"
```
Comment (bardielle) [inline pkg/operator/constants/constants.go:100]: @nirarg  I added that Cred key.
Please re-review

File: pkg/operator/metrics/metrics.go [hunk 5]
```diff
@@ -191,6 +191,8 @@ func cloudProviderSpecToMetricsKey(cloud string) string {
 		return "ovirt"
 	case "VsphereProviderSpec":
 		return "vsphere"
+	case "KubevirtProviderSpec":
+		return "kubevirt"
```
Comment (bardielle) [inline pkg/operator/metrics/metrics.go:197]: @nirarg  I added that new commit.
Please re-review

**Comment by rgolangh** [comment]
> `slog` or `a.getLogger`? use one

I just found out that I did the same mistake on ovirt actuator

**Comment by bardielle** [comment]
@rgolangh the usage of slog in createSecret (ocp/actuator) is a wrapping of the logger - Im using the a.gerlogger that I from the function update/ create 
I don't think it's a mistake  (also  I saw it in GCP)

**Comment by rgolangh** [comment]
> @rgolangh the usage of slog in createSecret (ocp/actuator) is a wrapping of the logger - Im using the a.gerlogger that I from the function update/ create
> I don't think it's a mistake (also I saw it in GCP)

I am going to fix that in my actuator. too many logger instances...

File: pkg/ocp/actuator.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+	logger.Debug("running delete")
+
+
+	existingSecret, err := a.getSecret(cr, logger)
+	if err != nil {
+		return err
+	}
+	if existingSecret != nil{
+		logger.Debug("Deleting existing secret")
+		err = a.Client.Delete(context.TODO(), existingSecret)
```
Comment (rgolangh) [inline pkg/ocp/actuator.go]: also here, pass `ctx`

File: pkg/ocp/actuator.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+			Annotations: map[string]string{
+				minterv1.AnnotationCredentialsRequest: fmt.Sprintf("%s/%s", cr.Namespace, cr.Name),
+			},
+		},
+		Data: map[string][]byte {
+			DataSecretKey : infraClusterKubeconfig,
+		},
+	}
+
+	err := a.Client.Create(context.TODO(), secret)
```
Comment (rgolangh) [inline pkg/ocp/actuator.go]: pass `ctx` , its the 1st argument of this method.

**Comment by bardielle** [comment]
@rgolangh slog and logger are the same instance log - slog is a wrapper that adds the filed "targetSecret" 
I went over all the other actuator - they used both of them 
In the create scenario (or Sync)  they used sLog and in all other used the logger

File: pkg/ocp/actuator.go [hunk 8]
```diff
@@ -121,47 +110,113 @@ func (a *OCPActuator) GetCredentialsRootSecretLocation() types.NamespacedName {
 	return types.NamespacedName{Namespace: constants.CloudCredSecretNamespace, Name: constants.KubevirtCloudCredSecretName}
 }
 
-func (a *OCPActuator) getSecret(ctx context.Context, cr *minterv1.CredentialsRequest, logger log.FieldLogger) (*corev1.Secret, error) {
-	logger.Debug("running Exists")
+func (a *OCPActuator) sync(ctx context.Context, cr *minterv1.CredentialsRequest,  logger log.FieldLogger) error {
+	logger.Debug("running sync")
```
Comment (nirarg) [inline pkg/ocp/actuator.go:113]: Remove this log, always printed log one line before (create/update)

File: pkg/ocp/actuator.go [hunk 9]
```diff
@@ -121,47 +110,113 @@ func (a *OCPActuator) GetCredentialsRootSecretLocation() types.NamespacedName {
...
-		return nil, err
+		logger.WithError(err).Error("issue with cloud credentials secret")
+		return err
 	}
 
-	logger.Debug("target secret exists")
-	return existingSecret, nil
-}
+	// get the existing secret in order to check if need to update or create a new
+	logger.Debugf("provisioning secret")
```
Comment (nirarg) [inline pkg/ocp/actuator.go]: Debugf is not needed here (Debug is enough) or is it a good point to print the kubevirtCredentialData?

File: pkg/ocp/actuator.go [hunk 10]
```diff
@@ -121,47 +110,113 @@ func (a *OCPActuator) GetCredentialsRootSecretLocation() types.NamespacedName {
...
+	}
 
-func (a *OCPActuator) createSecret(ctx context.Context, cr *minterv1.CredentialsRequest, logger log.FieldLogger) error {
-	sLog := logger.WithFields(log.Fields{
-		"targetSecret": fmt.Sprintf("%s/%s", cr.Spec.SecretRef.Namespace, cr.Spec.SecretRef.Name),
-		"cr":           fmt.Sprintf("%s/%s", cr.Namespace, cr.Name),
-	})
+	// check if need to update or create a new one
+	err = a.syncCredentialSecret(ctx, cr, &kubevirtCredentialData, existingSecret, logger)
+	if err != nil {
```
Comment (nirarg) [inline pkg/ocp/actuator.go]: One line

File: pkg/ocp/actuator.go [hunk 11]
```diff
@@ -121,47 +110,113 @@ func (a *OCPActuator) GetCredentialsRootSecretLocation() types.NamespacedName {
...
+	if kubevirtCredentialData != nil {
+		existingSecret.Data = map[string][]byte{
+			DataSecretKey: *kubevirtCredentialData,
+		}
+	}
+
+	if !reflect.DeepEqual(existingSecret, origSecret) {
+		sLog.Info("target secret has changed, updating")
+		err := a.Client.Update(context.TODO(), existingSecret)
+		if err != nil {
```
Comment (nirarg) [inline pkg/ocp/actuator.go]: One line

File: pkg/ocp/actuator.go [hunk 12]
```diff
@@ -121,47 +110,113 @@ func (a *OCPActuator) GetCredentialsRootSecretLocation() types.NamespacedName {
...
+			msg := "error updating secret"
+			sLog.WithError(err).Error(msg)
+			return &actuatoriface.ActuatorError{
+				ErrReason: minterv1.CredentialsProvisionFailure,
+				Message:   msg,
+			}
+		}
+	} else {
```
Comment (nirarg) [inline pkg/ocp/actuator.go:204]: else not needed because if do return

File: pkg/ocp/actuator.go [hunk 13]
```diff
@@ -181,10 +236,32 @@ func (a *OCPActuator) createSecret(ctx context.Context, cr *minterv1.Credentials
...
 	sLog.Info("secret created successfully")
 	return nil
 }
 
+func (a *OCPActuator) getSecret(ctx context.Context, cr *minterv1.CredentialsRequest, logger log.FieldLogger) (*corev1.Secret, error) {
+	logger.Debug("running Exists")
+
+	existingSecret := &corev1.Secret{}
+	err := a.Client.Get(ctx, types.NamespacedName{Namespace: cr.Spec.SecretRef.Namespace, Name: cr.Spec.SecretRef.Name}, existingSecret)
+	if err != nil {
```
Comment (nirarg) [inline pkg/ocp/actuator.go]: One line

File: pkg/ocp/actuator.go (hunk 8 above)
Comment (bardielle) [reply pkg/ocp/actuator.go:113]: It's a debug log so I think it will be nice to add that log

File: pkg/ocp/actuator.go (hunk 12 above)
Comment (bardielle) [reply pkg/ocp/actuator.go:204]: The else is on line 194 (if !reflect.DeepEqual(existingSecret, origSecret)) 
So it needed

File: pkg/ocp/actuator.go (hunk 8 above)
Comment (bardielle) [reply pkg/ocp/actuator.go:113]: It's only a debug log I think it will be nice for debugging

**Pull request closed without merging**